gps/deduce.go

```diff
@@ -167,36 +169,74 @@ func (m bitbucketDeducer) deduceRoot(path string) (string, error) {
 	return "bitbucket.org" + v[2], nil
 }
 
+// getBitbucketRepo expects the input to be the path after the bitbucket.org
+// hostname in an import URL. The following inputs should be valid:
+//
+// 		/user/repo
+// 		/user/repo.git
+// 		/user/repo/some/package
+// 		/user/repo.git/some/package
+func getBitbucketRepo(s string) (string, error) {
+	ss := strings.Split(s, "/")
+	if len(ss) < 3 {
+		return "", errors.Errorf("bitbucket repo path %q appears malformed, want /USER/REPO", s)
+	}
+
+	repoStr := ss[1] + "/" + ss[2]
+
+	if strings.HasSuffix(repoStr, ".git") {
+		repoStr = repoStr[:len(repoStr)-4]
+	}
+
+	if strings.HasSuffix(repoStr, ".hg") {
+		repoStr = repoStr[:len(repoStr)-3]
+	}
+
+	return repoStr, nil
+}
+
 func (m bitbucketDeducer) deduceSource(path string, u *url.URL) (maybeSources, error) {
 	v := m.regexp.FindStringSubmatch(path)
 	if v == nil {
 		return nil, fmt.Errorf("%s is not a valid path for a source on bitbucket.org", path)
 	}
 
+	bbRepo, err := getBitbucketRepo(v[2])
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to extract repo from bitbucket URL")
+	}
+
+	// the calls we are making are for repository metadata, which should be
+	// pretty quick.
+	//
+	// this timeout value is totally arbitrary, but should be within normal
+	// bounds
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+
+	scm, err := bitbucketapi.GetSCM(ctx, bbRepo)
+	cancel()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get SCM details from bitbucket repo")
+	}
+
+	isgit := scm == "git"
+
 	u.Host = "bitbucket.org"
```

Comment:
It doesn't seem super great to be manipulating the input?

Reply:
I guess this is already here

Reply:
I'm keeping the scope of this PR to only fix the things I know are broken. We can look at that, and the other thing you raised above about quoting the error string, but they are beyond the scope of this PR.

Reply:
ok